controllers/toolchainconfig/sync.go

```diff
@@ -0,0 +1,83 @@
+package toolchainconfig
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	"github.com/codeready-toolchain/toolchain-common/pkg/cluster"
+
+	"github.com/go-logr/logr"
+)
+
+type synchronizer struct {
+	getMembersFunc cluster.GetMemberClustersFunc
+	logger         logr.Logger
+}
+
+// syncMemberConfigs retrieves member operator configurations and syncs the appropriate configuration to each member cluster
+// returns a map of any errors encountered indexed by cluster name
+func (s *synchronizer) syncMemberConfigs(sourceConfig *toolchainv1alpha1.ToolchainConfig) map[string]string {
+	toolchainConfig := sourceConfig.DeepCopy()
+
+	// get configs for member toolchainclusters
+	memberToolchainClusters := s.getMembersFunc()
+	syncErrors := make(map[string]string, len(memberToolchainClusters))
+
+	if len(memberToolchainClusters) == 0 {
+		s.logger.Info("No toolchainclusters were found, skipping MemberOperatorConfig syncing")
+		return syncErrors
+	}
+
+	membersWithSpecificConfig := toolchainConfig.Spec.Members.SpecificPerMemberCluster
+
+	for _, toolchainCluster := range memberToolchainClusters {
+		memberConfigSpec := toolchainConfig.Spec.Members.Default
+
+		if c, ok := membersWithSpecificConfig[toolchainCluster.Name]; toolchainCluster.Type == cluster.Member && ok {
+			// member-specific configuration values override default configuration
+			memberConfigSpec = c
+			delete(membersWithSpecificConfig, toolchainCluster.Name)
+		}
+
+		if err := syncMemberConfig(memberConfigSpec, toolchainCluster); err != nil {
+			s.logger.Error(err, "failed to sync MemberOperatorConfig", "cluster_name", toolchainCluster.Name)
+			syncErrors[toolchainCluster.Name] = err.Error()
+		}
+	}
+
+	// add errors for any MemberOperatorConfigs that haven't been synced because there is no matching toolchaincluster
+	if len(membersWithSpecificConfig) > 0 {
+		for k := range membersWithSpecificConfig {
+			syncErrors[k] = "specific member configuration exists but no matching toolchaincluster was found"
+		}
+	}
```

Comment:
this returns an error for clusters that are not returned from `getMembersFunc()` but are specified in the ToolchainConfig. However, this code is not executed when no cluster is returned from `getMembersFunc()`. Shouldn't we remove the lines 30-33 https://github.com/codeready-toolchain/host-operator/blob/f38406bd2fb10c21cb9eb57d20e3f26c2cf32965/controllers/toolchainconfig/sync.go#L30-L33 ?

Reply:
Right, keep it as it is :+1: